src/EditorFeatures/CSharpTest/AddRequiredParentheses/AddRequiredParenthesesTests.cs

```diff
@@ -0,0 +1,434 @@
+    {
+        int x = 1 + 2 == 2 $$+ 3;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = 1 + 2 == (2 + 3);
+    }
+}", parameters: new TestParameters(options: RequireEqualityParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestCoalescePrecedence1()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    void M()
+    {
+        int x = a $$+ b ?? c;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = (a + b) ?? c;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestCoalescePrecedence2()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = a $$?? b ?? c;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestCoalescePrecedence3()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = a ?? b $$?? c;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestBitwisePrecedence1()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    void M()
+    {
+        int x = 1 $$+ 2 & 3;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = (1 + 2) & 3;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestBitwisePrecedence2()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = a $$| b | c;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestBitwisePrecedence3()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    void M()
+    {
+        int x = a | b $$& c;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int x = a | (b & c);
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddRequiredParentheses)]
+        public async Task TestBitwisePrecedence4()
+        {
+            await TestMissingAsync(
+@"class C
+{
+    void M()
+    {
+        int x = a $$| b & c;
+    }
+}", parameters: new TestParameters(options: RequireAllParenthesesForClarity));
+        }
+    }
+}
```

Comment:
Consider adding a test for `_ = a $$== b;` (action missing) and also some other operators (`+=`, also missing).

Reply:
```_ = a $$== b;``` Do you specifically want me to test ```_```?  Or was this just about == operators?

Reply:
sorry, things are super ambiguous here.  So if you could provide the explicit test cases i'd be happy to add them.

Reply:
- `bool b = a $$== b;` (I expect no action offered)
- `int x = x $$+= 5;` (I expect no action offered)
- `x $$= x.Next == null` (I'm not sure what I expect :smile: )

Reply:
Great.  will do.  thanks!

Reply:
Tests added.  #resolved.